File: examples/python/heleerra_api.py
"""
Heleerra Payment Gateway API Client

This module provides a Python client for integrating with the Heleerra payment gateway API.
Supports both sandbox and production environments with comprehensive error handling.
"""

import requests
import hmac
import hashlib
import json
import logging
from typing import Dict, Any, Optional, Union
from datetime import datetime
import time

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class HeleerraAPI:
    """
    Heleerra Payment Gateway API Client
    
    Provides methods for payment initiation, verification, and webhook handling.
    """

    def __init__(self, merchant_key: str, api_key: str, environment: str = 'sandbox', 
                 webhook_secret: Optional[str] = None, timeout: int = 30):
        """
        Initialize Heleerra API client

        Args:
            merchant_key (str): Your merchant ID from Heleerra dashboard
            api_key (str): Your API key from Heleerra dashboard
            environment (str): 'sandbox' or 'production'
            webhook_secret (str, optional): Webhook secret for signature verification
            timeout (int): Request timeout in seconds
        """
        self.base_url = 'https://heleerra.com'
        self.merchant_key = merchant_key
        self.api_key = api_key
        self.environment = environment.lower()
        self.webhook_secret = webhook_secret
        self.timeout = timeout
        
        # Set appropriate webhook secret based on environment
        if not webhook_secret:
            self.webhook_secret = f"test_webhook_secret" if self.environment == 'sandbox' else "webhook_secret"
        
        # Validate environment
        if self.environment not in ['sandbox', 'production']:
            raise ValueError("Environment must be either 'sandbox' or 'production'")
        
        logger.info(f"Heleerra API client initialized for {self.environment} environment")

    def initiate_payment(self, payment_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Initiate a new payment request

        Args:
            payment_data (dict): Payment details including amount, currency, etc.

        Returns:
            dict: API response with payment URL and transaction details

        Raises:
            ValueError: If required fields are missing or invalid
            requests.RequestException: If API request fails
        """
        self._validate_payment_data(payment_data)
        
        headers = {
            'Content-Type': 'application/json',
            'X-Environment': self.environment,
            'X-Merchant-Key': self.merchant_key,
            'X-API-Key': self.api_key
        }
        
        try:
            logger.info(f"Initiating payment for ref_trx: {payment_data.get('ref_trx', 'unknown')}")
            
            response = requests.post(
                f'{self.base_url}/api/v1/initiate-payment',
                json=payment_data,
                headers=headers,
                timeout=self.timeout
            )
            
            response.raise_for_status()
            
            result = response.json()
            logger.info("Payment initiated successfully", extra={
                'ref_trx': payment_data.get('ref_trx'),
                'amount': payment_data.get('payment_amount'),
                'environment': self.environment
            })
            
            return result
            
        except requests.HTTPError as e:
            error_data = response.json() if response.content else {}
            error_message = error_data.get('message', 'Payment initiation failed')
            error_code = error_data.get('error_code', 'UNKNOWN_ERROR')
            
            logger.error("Payment initiation failed", extra={
                'status_code': response.status_code,
                'error_code': error_code,
                'message': error_message,
                'ref_trx': payment_data.get('ref_trx')
            })
            
            raise requests.RequestException(f"Payment initiation failed: {error_message}")
            
        except requests.RequestException as e:
            logger.error("Network error during payment initiation", extra={
                'error': str(e),
                'ref_trx': payment_data.get('ref_trx')
            })
            raise

    def verify_payment(self, transaction_id: str) -> Dict[str, Any]:
        """
        Verify payment status

        Args:
            transaction_id (str): Heleerra transaction ID

        Returns:
            dict: Payment details and status

        Raises:
            requests.RequestException: If API request fails
        """
        headers = {
            'Accept': 'application/json',
            'X-Environment': self.environment,
            'X-Merchant-Key': self.merchant_key,
            'X-API-Key': self.api_key
        }
        
        try:
            logger.info(f"Verifying payment status for transaction: {transaction_id}")
            
            response = requests.get(
                f'{self.base_url}/api/v1/verify-payment/{transaction_id}',
                headers=headers,
                timeout=self.timeout
            )
            
            response.raise_for_status()
            
            result = response.json()
            logger.info("Payment verification successful", extra={
                'transaction_id': transaction_id,
                'status': result.get('status')
            })
            
            return result
            
        except requests.HTTPError as e:
            error_data = response.json() if response.content else {}
            error_message = error_data.get('message', 'Payment verification failed')
            
            logger.error("Payment verification failed", extra={
                'transaction_id': transaction_id,
                'status_code': response.status_code,
                'message': error_message
            })
            
            raise requests.RequestException(f"Payment verification failed: {error_message}")
            
        except requests.RequestException as e:
            logger.error("Network error during payment verification", extra={
                'transaction_id': transaction_id,
                'error': str(e)
            })
            raise

    def verify_webhook_signature(self, payload: str, signature: str) -> bool:
        """
        Verify webhook signature authenticity

        Args:
            payload (str): Raw webhook payload
            signature (str): Signature from X-Signature header

        Returns:
            bool: True if signature is valid
        """
        if not self.webhook_secret:
            logger.warning("Webhook secret not configured")
            return False

        try:
            expected_signature = hmac.new(
                self.webhook_secret.encode('utf-8'),
                payload.encode('utf-8'),
                hashlib.sha256
            ).hexdigest()
            
            # Use compare_digest to prevent timing attacks
            is_valid = hmac.compare_digest(expected_signature, signature)
            
            if not is_valid:
                logger.warning("Webhook signature verification failed")
            
            return is_valid
            
        except Exception as e:
            logger.error("Error verifying webhook signature", extra={'error': str(e)})
            return False

    def handle_webhook(self, payload: Dict[str, Any], headers: Dict[str, str]) -> Dict[str, Any]:
        """
        Handle webhook notification

        Args:
            payload (dict): Webhook payload
            headers (dict): Request headers

        Returns:
            dict: Response data
        """
        signature = headers.get('x-signature', '')
        environment = headers.get('x-environment', 'production')
        webhook_id = headers.get('x-webhook-id', '')

        # Verify signature
        if not self.verify_webhook_signature(json.dumps(payload), signature):
            logger.warning("Webhook signature verification failed", extra={
                'webhook_id': webhook_id,
                'environment': environment
            })
            
            return {
                'status': 'error',
                'message': 'Invalid signature'
            }

        # Handle based on environment
        if environment == 'sandbox':
            return self._handle_sandbox_webhook(payload)
        else:
            return self._handle_production_webhook(payload)

    def _handle_sandbox_webhook(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle sandbox webhook (testing environment)"""
        transaction_id = payload.get('data', {}).get('ref_trx')
        status = payload.get('status')
        amount = payload.get('data', {}).get('amount')

        logger.info("Processing sandbox webhook", extra={
            'transaction_id': transaction_id,
            'status': status,
            'amount': amount
        })

        # Sandbox-specific logic
        # Don't fulfill orders, don't send emails to real customers, etc.

        return {
            'status': 'sandbox_processed',
            'message': 'Webhook processed in sandbox mode'
        }

    def _handle_production_webhook(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle production webhook (live environment)"""
        status = payload.get('status', 'unknown')
        transaction_id = payload.get('data', {}).get('ref_trx')
        amount = payload.get('data', {}).get('amount', 0)

        logger.info("Processing production webhook", extra={
            'transaction_id': transaction_id,
            'status': status,
            'amount': amount
        })

        # Process based on payment status
        if status == 'completed':
            self._process_completed_payment(payload)
        elif status == 'failed':
            self._process_failed_payment(payload)
        elif status == 'cancelled':
            self._process_cancelled_payment(payload)
        elif status == 'expired':
            self._process_expired_payment(payload)
        else:
            logger.warning("Unknown payment status in webhook", extra={
                'status': status,
                'transaction_id': transaction_id
            })

        return {
            'status': 'processed',
            'message': 'Webhook processed successfully'
        }

    def _validate_payment_data(self, payment_data: Dict[str, Any]) -> None:
        """Validate payment data before sending to API"""
        required_fields = ['payment_amount', 'currency_code', 'ref_trx']
        
        for field in required_fields:
            if field not in payment_data or not payment_data[field]:
                raise ValueError(f"Required field '{field}' is missing or empty")

        amount = payment_data['payment_amount']
        if not isinstance(amount, (int, float)) or amount < 1.00:
            raise ValueError("Payment amount must be a number and at least 1.00")

        currency = payment_data['currency_code']
        if len(currency) != 3 or not currency.isalpha():
            raise ValueError("Currency code must be a 3-letter code (e.g., USD)")

        # Validate URLs if provided
        url_fields = ['success_redirect', 'failure_url', 'cancel_redirect', 'ipn_url']
        for field in url_fields:
            if field in payment_data and payment_data[field]:
                if not payment_data[field].startswith('http'):
                    raise ValueError(f"{field} must be a valid URL")

    def _process_completed_payment(self, payload: Dict[str, Any]) -> None:
        """Process completed payment"""
        transaction_id = payload.get('data', {}).get('ref_trx')
        amount = payload.get('data', {}).get('amount')
        customer_email = payload.get('data', {}).get('customer_email')

        # Update order status to completed
        # Send confirmation email to customer
        # Update inventory
        # Trigger any additional business logic

        logger.info("Payment completed successfully", extra={
            'transaction_id': transaction_id,
            'amount': amount,
            'customer_email': customer_email
        })

    def _process_failed_payment(self, payload: Dict[str, Any]) -> None:
        """Process failed payment"""
        transaction_id = payload.get('data', {}).get('ref_trx')
        
        # Update order status to failed
        # Send failure notification to customer
        # Log the failure for analysis

        logger.warning("Payment failed", extra={
            'transaction_id': transaction_id,
            'reason': payload.get('message', 'Unknown reason')
        })

    def _process_cancelled_payment(self, payload: Dict[str, Any]) -> None:
        """Process cancelled payment"""
        transaction_id = payload.get('data', {}).get('ref_trx')
        
        # Update order status to cancelled
        # Release any reserved inventory
        # Send cancellation notification

        logger.info("Payment cancelled by user", extra={
            'transaction_id': transaction_id
        })

    def _process_expired_payment(self, payload: Dict[str, Any]) -> None:
        """Process expired payment"""
        transaction_id = payload.get('data', {}).get('ref_trx')
        
        # Update order status to expired
        # Release any reserved inventory
        # Optionally send reminder to customer

        logger.info("Payment session expired", extra={
            'transaction_id': transaction_id
        })

    def get_config(self) -> Dict[str, Any]:
        """Get API configuration (for debugging)"""
        return {
            'environment': self.environment,
            'base_url': self.base_url,
            'merchant_key': self.merchant_key[:8] + '...',  # Mask for security
            'webhook_configured': bool(self.webhook_secret)
        }


class HeleerraAPIException(Exception):
    """Custom exception for Heleerra API errors"""
    pass


# Example usage
if __name__ == "__main__":
    # Initialize API client
    api = HeleerraAPI(
        merchant_key='test_merchant_123',
        api_key='test_api_key_abc',
        environment='sandbox'
    )

    # Example payment data
    payment_data = {
        'payment_amount': 250.00,
        'currency_code': 'USD',
        'ref_trx': f'ORDER_{int(time.time())}',
        'description': 'Premium Subscription',
        'success_redirect': 'https://yoursite.com/success',
        'failure_url': 'https://yoursite.com/failed',
        'cancel_redirect': 'https://yoursite.com/cancelled',
        'ipn_url': 'https://yoursite.com/webhooks/heleerra'
    }

    try:
        # Initiate payment
        result = api.initiate_payment(payment_data)
        print(f"Payment URL: {result['payment_url']}")
        
        # Verify payment (using example transaction ID)
        # verification_result = api.verify_payment('TXNQ5V8K2L9N3XM1')
        # print(f"Payment status: {verification_result['status']}")
        
    except Exception as e:
        print(f"Error: {e}")

File: examples/nodejs/heleerra_api.js
/**
 * Heleerra Payment Gateway API Client
 * 
 * This module provides a Node.js client for integrating with the Heleerra payment gateway API.
 * Supports both sandbox and production environments with comprehensive error handling.
 */

const axios = require('axios');
const crypto = require('crypto');

class HeleerraAPI {
    /**
     * Initialize Heleerra API client
     * 
     * @param {string} merchantKey - Your merchant ID from Heleerra dashboard
     * @param {string} apiKey - Your API key from Heleerra dashboard
     * @param {string} environment - 'sandbox' or 'production'
     * @param {string} webhookSecret - Webhook secret for signature verification
     * @param {number} timeout - Request timeout in milliseconds
     */
    constructor(merchantKey, apiKey, environment = 'sandbox', webhookSecret = null, timeout = 30000) {
        this.baseUrl = 'https://heleerra.com';
        this.merchantKey = merchantKey;
        this.apiKey = apiKey;
        this.environment = environment.toLowerCase();
        this.timeout = timeout;
        
        // Set appropriate webhook secret based on environment
        this.webhookSecret = webhookSecret || (this.environment === 'sandbox' ? 'test_webhook_secret' : 'webhook_secret');
        
        // Validate environment
        if (!['sandbox', 'production'].includes(this.environment)) {
            throw new Error("Environment must be either 'sandbox' or 'production'");
        }
        
        // Configure axios instance
        this.axios = axios.create({
            timeout: this.timeout,
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            }
        });
        
        // Add request logging
        this.axios.interceptors.request.use(
            (config) => {
                console.log(`Making ${config.method.toUpperCase()} request to ${config.url}`);
                return config;
            },
            (error) => {
                console.error('Request error:', error.message);
                return Promise.reject(error);
            }
        );
        
        // Add response logging
        this.axios.interceptors.response.use(
            (response) => {
                console.log(`Response received: ${response.status} ${response.statusText}`);
                return response;
            },
            (error) => {
                console.error('Response error:', error.response?.status, error.message);
                return Promise.reject(error);
            }
        );
        
        console.log(`Heleerra API client initialized for ${this.environment} environment`);
    }

    /**
     * Initiate a new payment request
     * 
     * @param {Object} paymentData - Payment details
     * @returns {Promise<Object>} API response with payment URL and transaction details
     */
    async initiatePayment(paymentData) {
        try {
            this._validatePaymentData(paymentData);
            
            const headers = {
                'X-Environment': this.environment,
                'X-Merchant-Key': this.merchantKey,
                'X-API-Key': this.apiKey
            };
            
            console.log(`Initiating payment for ref_trx: ${paymentData.ref_trx || 'unknown'}`);
            
            const response = await this.axios.post(
                `${this.baseUrl}/api/v1/initiate-payment`,
                paymentData,
                { headers }
            );
            
            const result = response.data;
            console.log('Payment initiated successfully', {
                ref_trx: paymentData.ref_trx,
                amount: paymentData.payment_amount,
                environment: this.environment
            });
            
            return result;
            
        } catch (error) {
            if (error.response) {
                const errorData = error.response.data || {};
                const errorMessage = errorData.message || 'Payment initiation failed';
                const errorCode = errorData.error_code || 'UNKNOWN_ERROR';
                
                console.error('Payment initiation failed', {
                    status: error.response.status,
                    error_code: errorCode,
                    message: errorMessage,
                    ref_trx: paymentData.ref_trx
                });
                
                throw new Error(`Payment initiation failed: ${errorMessage}`);
            } else {
                console.error('Network error during payment initiation', {
                    error: error.message,
                    ref_trx: paymentData.ref_trx
                });
                throw error;
            }
        }
    }

    /**
     * Verify payment status
     * 
     * @param {string} transactionId - Heleerra transaction ID
     * @returns {Promise<Object>} Payment details and status
     */
    async verifyPayment(transactionId) {
        try {
            const headers = {
                'Accept': 'application/json',
                'X-Environment': this.environment,
                'X-Merchant-Key': this.merchantKey,
                'X-API-Key': this.apiKey
            };
            
            console.log(`Verifying payment status for transaction: ${transactionId}`);
            
            const response = await this.axios.get(
                `${this.baseUrl}/api/v1/verify-payment/${transactionId}`,
                { headers }
            );
            
            const result = response.data;
            console.log('Payment verification successful', {
                transaction_id: transactionId,
                status: result.status
            });
            
            return result;
            
        } catch (error) {
            if (error.response) {
                const errorData = error.response.data || {};
                const errorMessage = errorData.message || 'Payment verification failed';
                
                console.error('Payment verification failed', {
                    transaction_id: transactionId,
                    status: error.response.status,
                    message: errorMessage
                });
                
                throw new Error(`Payment verification failed: ${errorMessage}`);
            } else {
                console.error('Network error during payment verification', {
                    transaction_id: transactionId,
                    error: error.message
                });
                throw error;
            }
        }
    }

    /**
     * Verify webhook signature authenticity
     * 
     * @param {string} payload - Raw webhook payload
     * @param {string} signature - Signature from X-Signature header
     * @returns {boolean} True if signature is valid
     */
    verifyWebhookSignature(payload, signature) {
        if (!this.webhookSecret) {
            console.warn('Webhook secret not configured');
            return false;
        }

        try {
            const expectedSignature = crypto
                .createHmac('sha256', this.webhookSecret)
                .update(payload, 'utf8')
                .digest('hex');
            
            const isValid = expectedSignature === signature;
            
            if (!isValid) {
                console.warn('Webhook signature verification failed');
            }
            
            return isValid;
            
        } catch (error) {
            console.error('Error verifying webhook signature', error);
            return false;
        }
    }

    /**
     * Handle webhook notification
     * 
     * @param {Object} payload - Webhook payload
     * @param {Object} headers - Request headers
     * @returns {Object} Response data
     */
    handleWebhook(payload, headers) {
        const signature = headers['x-signature'] || headers['X-Signature'] || '';
        const environment = headers['x-environment'] || headers['X-Environment'] || 'production';
        const webhookId = headers['x-webhook-id'] || headers['X-Webhook-ID'] || '';

        // Verify signature
        if (!this.verifyWebhookSignature(JSON.stringify(payload), signature)) {
            console.warn('Webhook signature verification failed', {
                webhook_id: webhookId,
                environment: environment
            });
            
            return {
                status: 'error',
                message: 'Invalid signature'
            };
        }

        // Handle based on environment
        if (environment === 'sandbox') {
            return this._handleSandboxWebhook(payload);
        } else {
            return this._handleProductionWebhook(payload);
        }
    }

    /**
     * Handle sandbox webhook (testing environment)
     * 
     * @param {Object} payload 
     * @returns {Object}
     */
    _handleSandboxWebhook(payload) {
        const transactionId = payload.data?.ref_trx;
        const status = payload.status;
        const amount = payload.data?.amount;

        console.log('Processing sandbox webhook', {
            transaction_id: transactionId,
            status: status,
            amount: amount
        });

        // Sandbox-specific logic
        // Don't fulfill orders, don't send emails to real customers, etc.

        return {
            status: 'sandbox_processed',
            message: 'Webhook processed in sandbox mode'
        };
    }

    /**
     * Handle production webhook (live environment)
     * 
     * @param {Object} payload 
     * @returns {Object}
     */
    _handleProductionWebhook(payload) {
        const status = payload.status || 'unknown';
        const transactionId = payload.data?.ref_trx;
        const amount = payload.data?.amount || 0;

        console.log('Processing production webhook', {
            transaction_id: transactionId,
            status: status,
            amount: amount
        });

        // Process based on payment status
        switch (status) {
            case 'completed':
                this._processCompletedPayment(payload);
                break;
            case 'failed':
                this._processFailedPayment(payload);
                break;
            case 'cancelled':
                this._processCancelledPayment(payload);
                break;
            case 'expired':
                this._processExpiredPayment(payload);
                break;
            default:
                console.warn('Unknown payment status in webhook', {
                    status: status,
                    transaction_id: transactionId
                });
        }

        return {
            status: 'processed',
            message: 'Webhook processed successfully'
        };
    }

    /**
     * Validate payment data before sending to API
     * 
     * @param {Object} paymentData 
     * @throws {Error}
     */
    _validatePaymentData(paymentData) {
        const required = ['payment_amount', 'currency_code', 'ref_trx'];
        
        for (const field of required) {
            if (!paymentData[field]) {
                throw new Error(`Required field '${field}' is missing or empty`);
            }
        }

        const amount = paymentData.payment_amount;
        if (typeof amount !== 'number' || amount < 1.00) {
            throw new Error('Payment amount must be a number and at least 1.00');
        }

        const currency = paymentData.currency_code;
        if (currency.length !== 3) {
            throw new Error('Currency code must be a 3-letter code (e.g., USD)');
        }

        // Validate URLs if provided
        const urlFields = ['success_redirect', 'failure_url', 'cancel_redirect', 'ipn_url'];
        for (const field of urlFields) {
            if (paymentData[field] && !paymentData[field].startsWith('http')) {
                throw new Error(`${field} must be a valid URL`);
            }
        }
    }

    /**
     * Process completed payment
     * 
     * @param {Object} payload 
     */
    _processCompletedPayment(payload) {
        const transactionId = payload.data?.ref_trx;
        const amount = payload.data?.amount;
        const customerEmail = payload.data?.customer_email;

        // Update order status to completed
        // Send confirmation email to customer
        // Update inventory
        // Trigger any additional business logic

        console.log('Payment completed successfully', {
            transaction_id: transactionId,
            amount: amount,
            customer_email: customerEmail
        });
    }

    /**
     * Process failed payment
     * 
     * @param {Object} payload 
     */
    _processFailedPayment(payload) {
        const transactionId = payload.data?.ref_trx;
        
        // Update order status to failed
        // Send failure notification to customer
        // Log the failure for analysis

        console.warn('Payment failed', {
            transaction_id: transactionId,
            reason: payload.message || 'Unknown reason'
        });
    }

    /**
     * Process cancelled payment
     * 
     * @param {Object} payload 
     */
    _processCancelledPayment(payload) {
        const transactionId = payload.data?.ref_trx;
        
        // Update order status to cancelled
        // Release any reserved inventory
        // Send cancellation notification

        console.log('Payment cancelled by user', {
            transaction_id: transactionId
        });
    }

    /**
     * Process expired payment
     * 
     * @param {Object} payload 
     */
    _processExpiredPayment(payload) {
        const transactionId = payload.data?.ref_trx;
        
        // Update order status to expired
        // Release any reserved inventory
        // Optionally send reminder to customer

        console.log('Payment session expired', {
            transaction_id: transactionId
        });
    }

    /**
     * Get API configuration (for debugging)
     * 
     * @returns {Object}
     */
    getConfig() {
        return {
            environment: this.environment,
            base_url: this.baseUrl,
            merchant_key: this.merchantKey.substring(0, 8) + '...', // Mask for security
            webhook_configured: !!this.webhookSecret
        };
    }
}

// Example usage
if (require.main === module) {
    // Initialize API client
    const api = new HeleerraAPI(
        'test_merchant_123',
        'test_api_key_abc',
        'sandbox'
    );

    // Example payment data
    const paymentData = {
        payment_amount: 250.00,
        currency_code: 'USD',
        ref_trx: `ORDER_${Date.now()}`,
        description: 'Premium Subscription',
        success_redirect: 'https://yoursite.com/success',
        failure_url: 'https://yoursite.com/failed',
        cancel_redirect: 'https://yoursite.com/cancelled',
        ipn_url: 'https://yoursite.com/webhooks/heleerra'
    };

    // Example usage
    api.initiatePayment(paymentData)
        .then(result => {
            console.log('Payment URL:', result.payment_url);
            
            // Example: Verify payment
            // return api.verifyPayment('TXNQ5V8K2L9N3XM1');
        })
        // .then(verificationResult => {
        //     console.log('Payment status:', verificationResult.status);
        // })
        .catch(error => {
            console.error('Error:', error.message);
        });
}

module.exports = HeleerraAPI;